core/src/impl/Kokkos_AnalyzePolicy.hpp

```diff
@@ -250,14 +296,47 @@ struct AnalyzePolicy<Base> {
 
   using is_graph_kernel = typename Base::is_graph_kernel;
 
-  using type = PolicyTraitsBase<execution_space, schedule_type, work_tag,
-                                index_type, iteration_pattern, launch_bounds,
-                                work_item_property, is_graph_kernel>;
+  using occupancy_control = typename Base::occupancy_control;
+
+  using type =
+      PolicyTraitsBase<execution_space, schedule_type, work_tag, index_type,
+                       iteration_pattern, launch_bounds, work_item_property,
+                       is_graph_kernel, occupancy_control>;
+};
+
+template <bool Enable = false>
+struct MaybeDesiredOccupancy {};
+
+template <>
+struct MaybeDesiredOccupancy<true> {
+  Kokkos::Experimental::DesiredOccupancy m_occupancy;
 };
 
 template <typename... Traits>
 struct PolicyTraits
-    : public AnalyzePolicy<PolicyTraitsBase<>, Traits...>::type {};
+    : public AnalyzePolicy<PolicyTraitsBase<>, Traits...>::type,
+      private MaybeDesiredOccupancy<std::is_same<
+          typename AnalyzePolicy<PolicyTraitsBase<>,
+                                 Traits...>::type::occupancy_control,
```

Comment:
Non-blocking: I don't like repeating this here, but I guess when we rework the policy properties design we can fix this.